### PR TITLE
Draft: Added - Simple drop down Quality Picker for HLS & Dash players (no MSS yet)

### PR DESCRIPTION
**Pull request by Nfrederiksen**
Hej hej fellow developers.

I am here to learn from my misstakes, and slowly but surely chistle away on my dev skills.
Please share your knowledge.

Started with a drop down menu as quality picker interface, with support for dash & hls. I reckon that ideally, when it works 100% should be added as an icon-button perhaps?

Issues: I am stuck, can't see why live HLS streams won't load a level that is not -1/auto-select. There is likely more to it. I'm happy to learn.

webp_img

### CODE REVIEW

File: packages/core/src/tech/HlsJsTech.ts [hunk 1]
```diff
@@ -75,6 +75,39 @@ export default class HlsJsTech extends BaseTech {
     this.playlistDuration = data?.details?.totalduration;
   }
 
+  getVideoQualities(){
+    if(this.hls){
+      const videoQualities: IVideoQuality[] = [];
+      const levels = this.hls.levels;
```
Comment (bwallberg) [inline packages/core/src/tech/HlsJsTech.ts:81]: You can optimally just return a mapped array instead `return this.hls.levels.map(...)` ( https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map )

File: packages/core/src/tech/HlsJsTech.ts [hunk 2]
```diff
@@ -75,6 +75,39 @@ export default class HlsJsTech extends BaseTech {
...
+        videoQualities.push(quality);
+      });
+      return videoQualities;
+    }
+  }
+
+  get currentLevel(){
+    if(this.hls)
+      return this.hls.currentLevel;
+    return NaN;
```
Comment (bwallberg) [inline packages/core/src/tech/HlsJsTech.ts:97]: `NaN` is generally considered bad practise, the function will automatically return `undefined` without it which is fine.

Comment (Nfrederiksen) [reply packages/core/src/tech/HlsJsTech.ts:97]: yessir

File: packages/core/src/tech/HlsJsTech.ts (hunk 1 above)
Comment (Nfrederiksen) [reply packages/core/src/tech/HlsJsTech.ts:81]: Alright. great tip.

Comment (Nfrederiksen) [reply packages/core/src/tech/HlsJsTech.ts]: BTW... any insight as to why `this.hls.currentLevel = level`, freezes the player when `level` is !=-1, for when playing this kind of hls stream ? 😖 
https://cph-p2p-msl.akamaized.net/hls/live/2000341/test/master.m3u8

**Review by bwallberg — CHANGES_REQUESTED** [review]
I think it looks pretty good! I have two general notes:

- [Opinion] I think we should create a general "debug options" area _below_ the player, if we add multiple options above the player we might have to start scrolling to see the player
- You should run the files you've modified through prettier ( https://prettier.io/ ), there are multiple plugins available for different editors for this, in the future we'll add a commit hook so you can commit files that haven't gone through prettier.

File: packages/demo/src/index.js [hunk 3]
```diff
@@ -39,12 +41,30 @@ async function main() {
...
-  function load() {
-    player.load(manifestInput.value);
+  async function load() {
+    await player.load(manifestInput.value);
+    populateQualityPicker();
   }
 
+  function populateQualityPicker(){
+    // Reset/Clear-out the drop down menu.
+    while(qualityPicker.options.length > 1){
```
Comment (bwallberg) [inline packages/demo/src/index.js:53]: I think a bit of a safer variant would be to loop through the options and remove the options without the value `-1`, theoretically you could end up removing auto as is.

File: packages/demo/src/index.js [hunk 4]
```diff
@@ -39,12 +41,30 @@ async function main() {
...
+      qualityPicker.remove(1);
+     }
+
+    const videoQualities = player.getAllVideoQualities();
+    videoQualities.forEach((element, level) => {
+      const option = document.createElement('option');
+      option.text = `${element['width']}x${element['height']}, ${Math.round(element['bitrate'] / 1024)}kbps`;
```
Comment (bwallberg) [inline packages/demo/src/index.js:60]: 👌

File: packages/demo/src/index.js (hunk 3 above)
Comment (Nfrederiksen) [reply packages/demo/src/index.js:53]: I see, in case a value ever pushes -1 beyond index 0. 🤔

Comment (bwallberg) [reply packages/demo/src/index.js]: Yes, exactly. Which, of course, isn't likely to ever happen but by being explicit with your intention you are both making the code a bit safer and easier to read :)

File: packages/demo/src/index.js [hunk 5]
```diff
@@ -39,12 +39,31 @@ async function main() {
...
   }
 
+  function populateQualityPicker() {
+    // Reset/Clear-out the drop down menu.
+    for (var i = qualityPicker.options.length - 1; i >= 0; i--) {
+      if (qualityPicker.options[i].value != '-1') qualityPicker.remove(i);
+    }
+
+    const videoQualities = player.getAllVideoQualities();
+    videoQualities.forEach((element, level) => {
```
Comment (bwallberg) [inline packages/demo/src/index.js:56]: The IVideoQuality type should have the `id` as part of the object instead, there is no guarantee that the Ids are numbers for different types of playback engines.

File: packages/core/src/tech/HlsJsTech.ts (hunk 1 above)
Comment (bwallberg) [reply packages/core/src/tech/HlsJsTech.ts]: @Nfrederiksen
It seems something is wonky with that stream, not sure what and I don't have time to dig further unfortunetely. But your implementation is fine so it's not that.

File: packages/core/src/tech/HlsJsTech.ts [hunk 6]
```diff
@@ -75,6 +79,34 @@ export default class HlsJsTech extends BaseTech {
     this.playlistDuration = data?.details?.totalduration;
   }
 
+  getVideoQualities() {
```
Comment (bwallberg) [inline packages/core/src/tech/HlsJsTech.ts:82]: For naming consistency I'd rather call this `getVideoLevels`

File: packages/core/src/tech/HlsJsTech.ts [hunk 7]
```diff
@@ -75,6 +79,34 @@ export default class HlsJsTech extends BaseTech {
...
+          height: level['height'],
+          bitrate: level['bitrate'],
+        };
+        return quality;
+      });
+    }
+  }
+
+  get currentLevel() {
+    if (this.hls) return this.hls.currentLevel;
```
Comment (bwallberg) [inline packages/core/src/tech/HlsJsTech.ts:96]: Please return the full level instead, basically the `IVideoQuality` item ( that should be renamed `IVideoLevel` )

File: packages/core/src/tech/HlsJsTech.ts [hunk 8]
```diff
@@ -75,6 +79,34 @@ export default class HlsJsTech extends BaseTech {
...
+    return -1;
+  }
+
+  set currentLevel(level: number) {
```
Comment (bwallberg) [inline packages/core/src/tech/HlsJsTech.ts:100]: Set should be called with the full `IVIdeoLevel` instead.

File: packages/demo/src/index.js (hunk 5 above)
Comment (Nfrederiksen) [reply packages/demo/src/index.js:56]: Ah. Okej 👍

File: packages/demo/src/index.js [hunk 9]
```diff
@@ -39,12 +39,31 @@ async function main() {
...
+      const option = document.createElement('option');
+      option.text = `${level.width}x${level.height}, ${Math.round(
+        level.bitrate / 1024
+      )}kbps`;
+      option.value = index;
+      qualityPicker.add(option);
+    });
+  }
+  // Quality picker fails with:
+  // https://cph-p2p-msl.akamaized.net/hls/live/2000341/test/master.m3u8
```
Comment (bwallberg) [inline packages/demo/src/index.js:66]: remove this dev comment

**Pull request closed without merging**